clang/lib/Sema/SemaExpr.cpp

```diff
@@ -12186,13 +12186,39 @@ Sema::CheckValueTerminatedAssignmentConstraints(QualType LHSType,
              Context.hasSameUnqualifiedType(
                  Context.getCorrespondingSignedType(LPointee),
                  Context.getCorrespondingSignedType(RPointee)));
-        if (CompatiblePointees &&
+        if (CompatiblePointees && RHSExpr->isPRValue() &&
             RHSExpr->tryEvaluateTerminatorElement(Evald, Context) &&
             Evald.Val.isInt() &&
             llvm::APSInt::isSameValue(LVTT->getTerminatorValue(Context),
                                       Evald.Val.getInt())) {
           return Sema::Compatible;
         }
+        // If in C++ Safe Buffers/Bounds Safety interoperation mode, the RHS can
+        // be 'std::string::c_str':
+        bool SafeBuffersEnabled = !Diags.isIgnored(
+            diag::warn_unsafe_assign_to_null_terminated_pointer,
+            RHSExpr->getBeginLoc());
+
+        if (LangOpts.CPlusPlus && SafeBuffersEnabled) {
+          if (const auto *MCE =
+                  dyn_cast<CXXMemberCallExpr>(RHSExpr->IgnoreParenImpCasts())) {
+            IdentifierInfo *MethodDeclII = nullptr, *ObjTypeII = nullptr;
+
+            if (MCE->getMethodDecl())
+              MethodDeclII = MCE->getMethodDecl()->getIdentifier();
+            if (const auto *RecordDecl =
+                    MCE->getObjectType()->getAsRecordDecl()) {
+              // If not in std namespace, let `ObjTypeII` be null so that the
+              // match fails:
+              if (RecordDecl->isInStdNamespace())
+                ObjTypeII = RecordDecl->getIdentifier();
+            }
+            if (MethodDeclII && ObjTypeII &&
```

Comment:
I think this could be combined with the prior if-statement for checking that RecordDecl is in the standard namespace. It would clearer.